sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,757 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Messaging.EventHubs.Core;
+using Azure.Messaging.EventHubs.Diagnostics;
+
+namespace Azure.Messaging.EventHubs.Processor
+{
+    /// <summary>
+    ///   An abstraction used as a base for the <see cref="EventProcessorClient" /> class.  It provides partition processing load
+    ///   balancing across multiple processor instances in the context of an Event Hub and a consumer group, and it can be extended
+    ///   by a concrete implementation with customized functionality.
+    /// </summary>
+    ///
+    /// <typeparam name="T">An underlying type with additional information that fully describes the context in which a partition is being processed.</typeparam>
+    ///
+    public abstract class EventProcessorBase<T>
+    {
+        /// <summary>The seed to use for initializing random number generated for a given thread-specific instance.</summary>
+        private static int s_randomSeed = Environment.TickCount;
```

Comment:
It seems an interesting approach. Certainly after-release.

Load balancing basically consists of a bunch of small tasks that may run under certain conditions. I think separating these in multiple methods may be simple to do, and would in fact make testability easier. However, I'm not sure to what extent would these various methods calls impact the algorithm performance.

Reply:
I was thinking something more comprehensive plug-and-play, where all of the load balancing was abstracted into an object that we just tell "do the load balancing" and that has some properties to signal what's going on.     With the new design, this will make even more sense because our Blobs implementation and our base implementation will likely need that abstraction to share the logic.

But.... yeah, post-release for sure.